pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,353 @@
+		cfg:        cfg,
+		queue:      queue,
+		privateKey: privateKey,
+	}
+}
+
+// Name implements the core.Service interface.
+func (bfs *Service) Name() string {
+	return "BlockFetcherService"
+}
+
+// Start implements the core.Service interface.
+func (bfs *Service) Start() {
+	bfs.log.Info("Starting Block Fetcher Service")
+	c, err := neofs.GetSDKClient(context.Background(), bfs.cfg.Addresses[0])
+	defer func(c *client.Client) {
+		err := c.Close()
+		if err != nil {
+			bfs.log.Error("Failed to close SDK client", zap.Error(err))
+		}
+	}(c)
+
+	if err != nil {
+		bfs.log.Error("Failed to create SDK client", zap.Error(err))
+		close(bfs.quit)
+		return
+	}
+	bfs.client = c
+	switch bfs.cfg.Mode {
+	case ModeIndexSearch:
+		err = bfs.fetchData()
+	case ModeOidSearch:
+		err = bfs.fetchDataWithOid()
+	default:
+		bfs.log.Error("Invalid mode specified", zap.String("mode", bfs.cfg.Mode))
+		close(bfs.quit)
+		return
+	}
+
+	if err != nil {
+		bfs.log.Error("Fetch operation failed", zap.Error(err))
+		close(bfs.quit)
+		return
+	}
+}
+
+func (bfs *Service) GetHeaders() error {
+	startIndex := bfs.chain.HeaderHeight() / 2000
+
+	for {
+		select {
+		case <-bfs.quit:
+			return nil
+		default:
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter("index_header", fmt.Sprintf("%d", startIndex), object.MatchNumGE)
+			filters.AddFilter("index_header", fmt.Sprintf("%d", startIndex), object.MatchNumLE)
+			prm.SetFilters(filters)
+
+			headerOids, err := bfs.search(prm)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to find 'index_header' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			if len(headerOids) == 0 {
+				bfs.log.Info(fmt.Sprintf("No 'index_header' object found with index %d, stopping.", startIndex))
+				return errors.New("no 'index_header' object found")
+			}
+
+			data, err := bfs.get(headerOids[0].String())
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to fetch 'index_header' object with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			err = bfs.processHeaders(data)
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to process headers for index %d", startIndex), zap.Error(err))
+				return err
+			}
+			startIndex++
+		}
+	}
+}
+
+// Shutdown implements the core.Service interface.
+func (bfs *Service) Shutdown() {
+	bfs.log.Info("Shutting down Block Fetcher Service")
+	close(bfs.quit)
+}
+
+// IsActive implements the Service interface.
+func (bfs *Service) IsActive() bool {
+	if !bfs.cfg.Enabled {
+		return false
+	}
+	select {
+	case <-bfs.quit:
+		return false
+	default:
+		return true
+	}
+}
+
+func (bfs *Service) fetchData() error {
+	startIndex := bfs.chain.BlockHeight()
+	batchSize := uint32(30)
+
+	for {
+		select {
+		case <-bfs.quit:
+			bfs.log.Info("Stopping fetchData operation.")
+			return nil
+		default:
+			bfs.log.Info(fmt.Sprintf("Current BlockHeight %d", bfs.chain.BlockHeight()))
```

Comment:
It's too chatty, there's Blockchain's persist log and thus we don't need this log.